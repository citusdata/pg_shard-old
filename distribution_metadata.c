/*-------------------------------------------------------------------------
 *
 * distribution_metadata.c
 *		  Cluster metadata handling for pg_shard
 *
 * Portions Copyright (c) 2014, Citus Data, Inc.
 *
 * IDENTIFICATION
 *		  shard_metadata.c
 *
 *-------------------------------------------------------------------------
 */

#include "postgres.h"
#include "fmgr.h"

#include "distribution_metadata.h"

#include "access/genam.h"
#include "access/heapam.h"
#include "access/htup_details.h"
#include "access/htup.h"
#include "access/skey.h"
#include "catalog/pg_type.h"
#include "executor/executor.h"
#include "nodes/makefuncs.h"
#include "nodes/params.h"
#include "nodes/pg_list.h"
#include "nodes/primnodes.h"
#include "parser/parse_node.h"
#include "parser/parse_relation.h"
#include "utils/builtins.h"
#include "utils/elog.h"
#include "utils/fmgroids.h"
#include "utils/lsyscache.h"
#include "utils/palloc.h"
#include "utils/rel.h"
#include "utils/tqual.h"


/* local function forward declarations */
static Var * ColumnNameToVar(Relation relation, char *columnName);
static void LoadShardRow(int64 shardId, Oid *relationId, char **minValue,
						 char **maxValue);
static Placement * TupleToPlacement(HeapTuple heapTuple,
									TupleDesc tupleDescriptor);


/* declarations for dynamic loading */
PG_FUNCTION_INFO_V1(TestDistributionMetadata);


/*
 * TestDistributionMetadata prints all current shard and placement configuration
 * at INFO level for testing purposes.
 */
Datum
TestDistributionMetadata(PG_FUNCTION_ARGS)
{
	Oid distributedTableId = PG_GETARG_OID(0);

	Var *partitionColumn = PartitionColumn(distributedTableId);
	List *shardList = LoadShardList(distributedTableId);
	List *placementList = NIL;

	ListCell *cell = NULL;

	FmgrInfo outputFunctionInfo = { };
	Oid outputFunctionId = InvalidOid;
	bool isVarlena = false;

	/* then find min/max values' actual types */
	getTypeOutputInfo(partitionColumn->vartype, &outputFunctionId, &isVarlena);
	fmgr_info(outputFunctionId, &outputFunctionInfo);

	ereport(INFO, (errmsg("Table is partitioned using column #%d, "
						  "which is of type \"%s\"", partitionColumn->varattno,
						  format_type_be(partitionColumn->vartype))));

	ereport(INFO, (errmsg("Found %d shards...", list_length(shardList))));

	foreach(cell, shardList)
	{
		ListCell *placementCell = NULL;
		int64 *shardId = NULL;

		shardId = (int64 *) lfirst(cell);
		Shard * shard = LoadShard(*shardId);

		char *minValueStr = OutputFunctionCall(&outputFunctionInfo, shard->minValue);
		char *maxValueStr = OutputFunctionCall(&outputFunctionInfo, shard->maxValue);

		ereport(INFO, (errmsg("Shard #" INT64_FORMAT, shard->id)));
		ereport(INFO, (errmsg("\trelation:\t%s",
							  get_rel_name(shard->relationId))));

		ereport(INFO, (errmsg("\tmin value:\t%s", minValueStr)));
		ereport(INFO, (errmsg("\tmax value:\t%s", maxValueStr)));

		placementList = LoadPlacementList(*shardId);
		ereport(INFO, (errmsg("\t%d placements:",
							  list_length(placementList))));

		foreach(placementCell, placementList)

		{
			Placement *placement = NULL;

			placement = (Placement *) lfirst(placementCell);

			ereport(INFO, (errmsg("\t\tPlacement #" INT64_FORMAT,
								  placement->id)));
			ereport(INFO, (errmsg("\t\t\tshard:\t" INT64_FORMAT,
								  placement->shardId)));
			ereport(INFO, (errmsg("\t\t\tnode name:\t%s",
								  placement->nodeName)));
			ereport(INFO, (errmsg("\t\t\tnode port:\t%u",
								  placement->nodePort)));
		}
	}

	PG_RETURN_VOID();
}


/*
 * LoadShardList returns a List of shard identifiers related for a given
 * distributed table. If no shards can be found for the specified relation, an
 * empty List is returned.
 */
List *
LoadShardList(Oid distributedTableId)
{
	const int scanKeyCount = 1;

	List *shardList = NIL;

	RangeVar *heapRangeVar = NULL, *indexRangeVar = NULL;
	Relation heapRelation = NULL, indexRelation = NULL;
	IndexScanDesc indexScanDesc = NULL;
	ScanKeyData scanKey[scanKeyCount];
	HeapTuple heapTuple = NULL;
	bool isNull = false;

	heapRangeVar = makeRangeVar(METADATA_SCHEMA, SHARD_TABLE_NAME, -1);
	indexRangeVar = makeRangeVar(METADATA_SCHEMA, SHARD_RELATION_IDX, -1);

	heapRelation = relation_openrv(heapRangeVar, AccessShareLock);
	indexRelation = relation_openrv(indexRangeVar, AccessShareLock);

	ScanKeyInit(&scanKey[0], 1, BTEqualStrategyNumber, F_OIDEQ,
				ObjectIdGetDatum(distributedTableId));

	indexScanDesc = index_beginscan(heapRelation, indexRelation, SnapshotNow,
								  scanKeyCount, 0);
	index_rescan(indexScanDesc, scanKey, scanKeyCount, NULL, 0);

	// TODO: Do I need to check scan->xs_recheck and recheck scan key?
	heapTuple = index_getnext(indexScanDesc, ForwardScanDirection);
	while (HeapTupleIsValid(heapTuple))
	{
		TupleDesc tupleDescriptor = RelationGetDescr(heapRelation);
		Datum shardIdDatum = heap_getattr(heapTuple, ATTR_NUM_SHARD_ID,
										  tupleDescriptor, &isNull);

		int64 shardId = DatumGetInt64(shardIdDatum);
		int64 *shardIdPointer = (int64 *) palloc0(sizeof(int64));
		*shardIdPointer = shardId;

		shardList = lappend(shardList, shardIdPointer);

		heapTuple = index_getnext(indexScanDesc, ForwardScanDirection);
	}

	index_endscan(indexScanDesc);
	index_close(indexRelation, AccessShareLock);
	relation_close(heapRelation, AccessShareLock);

	return shardList;
}


/*
 * LoadShard collects metadata for a specified shard in a Shard and returns a
 * pointer to that structure. If no shard can be found using the provided
 * identifier, an error is thrown.
 */
Shard *
LoadShard(int64 shardId)
{
	Shard *shard = NULL;
	Datum minValue = 0;
	Datum maxValue = 0;
	Var *partitionColumn = NULL;
	Oid inputFunctionId = InvalidOid;
	Oid typeIoParam = InvalidOid;
	Oid relationId = InvalidOid;
	FmgrInfo inputFunctionInfo = { };
	char *minValueString = NULL;
	char *maxValueString = NULL;

	/* first read the related row from the shard table */
	LoadShardRow(shardId, &relationId, &minValueString, &maxValueString);

	/* then find min/max values' actual types */
	partitionColumn = PartitionColumn(relationId);
	getTypeInputInfo(partitionColumn->vartype, &inputFunctionId, &typeIoParam);
	fmgr_info(inputFunctionId, &inputFunctionInfo);

	/* finally convert min/max values to their actual types */
	minValue = InputFunctionCall(&inputFunctionInfo, minValueString,
								 typeIoParam, partitionColumn->vartypmod);
	maxValue = InputFunctionCall(&inputFunctionInfo, maxValueString,
								 typeIoParam, partitionColumn->vartypmod);

	shard = (Shard *) palloc0(sizeof(Shard));
	shard->id = shardId;
	shard->relationId = relationId;
	shard->minValue = minValue;
	shard->maxValue = maxValue;
	shard->valueTypeId = INT4OID;	// we only deal with hash ranges for now

	return shard;
}


/*
 * LoadPlacementList gathers placement metadata for every placement of a given
 * shard and returns a List of Placements containing that metadata. If no
 * placements exist for the specified shard, an empty list is returned.
 */
List *
LoadPlacementList(int64 shardId)
{
	const int scanKeyCount = 1;

	List *placementList = NIL;

	RangeVar *heapRangeVar = NULL, *indexRangeVar = NULL;
	Relation heapRelation = NULL, indexRelation = NULL;
	IndexScanDesc indexScanDesc = NULL;
	ScanKeyData scanKey[scanKeyCount];
	HeapTuple heapTuple = NULL;

	heapRangeVar = makeRangeVar(METADATA_SCHEMA, PLACEMENT_TABLE_NAME, -1);
	indexRangeVar = makeRangeVar(METADATA_SCHEMA, PLACEMENT_SHARD_IDX, -1);

	heapRelation = relation_openrv(heapRangeVar, AccessShareLock);
	indexRelation = relation_openrv(indexRangeVar, AccessShareLock);

	ScanKeyInit(&scanKey[0], 1, BTEqualStrategyNumber, F_INT8EQ,
				Int64GetDatum(shardId));

	indexScanDesc = index_beginscan(heapRelation, indexRelation, SnapshotNow,
								  scanKeyCount, 0);
	index_rescan(indexScanDesc, scanKey, scanKeyCount, NULL, 0);

	// TODO: Do I need to check scan->xs_recheck and recheck scan key?
	heapTuple = index_getnext(indexScanDesc, ForwardScanDirection);
	while (HeapTupleIsValid(heapTuple))
	{
		TupleDesc tupleDescriptor = RelationGetDescr(heapRelation);
		Placement *placement = TupleToPlacement(heapTuple, tupleDescriptor);
		placementList = lappend(placementList, placement);

		heapTuple = index_getnext(indexScanDesc, ForwardScanDirection);
	}

	index_endscan(indexScanDesc);
	index_close(indexRelation, AccessShareLock);
	relation_close(heapRelation, AccessShareLock);

	/* if no shard placements are found, warn the user */
	if (placementList == NIL)
	{
		ereport(WARNING, (errmsg("could not find any placements for shardId "
								 INT64_FORMAT, shardId)));
	}

	return placementList;
}


/*
 * PartitionColumn looks up the column used to partition a given distributed
 * table and returns a reference to a Var representing that column. If no entry
 * can be found using the provided identifer, an error is thrown.
 */
Var *
PartitionColumn(Oid distributedTableId)
{
	const int scanKeyCount = 1;

	RangeVar *heapRangeVar = NULL, *indexRangeVar = NULL;
	Relation heapRelation = NULL, indexRelation = NULL;
	IndexScanDesc indexScanDesc = NULL;
	ScanKeyData scanKey[scanKeyCount];
	HeapTuple heapTuple = NULL;

	Var *partitionColumn = NULL;

	heapRangeVar = makeRangeVar(METADATA_SCHEMA,
								PARTITION_STRATEGY_TABLE_NAME, -1);
	indexRangeVar = makeRangeVar(METADATA_SCHEMA,
								 PARTITION_STRATEGY_RELATION_IDX, -1);

	heapRelation = relation_openrv(heapRangeVar, AccessShareLock);
	indexRelation = relation_openrv(indexRangeVar, AccessShareLock);

	ScanKeyInit(&scanKey[0], 1, BTEqualStrategyNumber, F_OIDEQ,
				ObjectIdGetDatum(distributedTableId));

	indexScanDesc = index_beginscan(heapRelation, indexRelation, SnapshotNow,
								  scanKeyCount, 0);
	index_rescan(indexScanDesc, scanKey, scanKeyCount, NULL, 0);

	// TODO: Do I need to check scan->xs_recheck and recheck scan key?
	heapTuple = index_getnext(indexScanDesc, ForwardScanDirection);
	if (HeapTupleIsValid(heapTuple))
	{
		bool isNull = false;
		TupleDesc tupleDescriptor = RelationGetDescr(heapRelation);

		Datum keyDatum = heap_getattr(heapTuple,
									  ATTR_NUM_PARTITION_STRATEGY_KEY,
									  tupleDescriptor, &isNull);
		char *partitionColumnName = TextDatumGetCString(keyDatum);

		Relation relation = relation_open(distributedTableId,
										  AccessShareLock);
		partitionColumn = ColumnNameToVar(relation, partitionColumnName);
		relation_close(relation, AccessShareLock);
	}
	else
	{
		ereport(ERROR, (errmsg("could not find partition for distributed"
							   "relation %u", distributedTableId)));
	}

	index_endscan(indexScanDesc);
	index_close(indexRelation, AccessShareLock);
	relation_close(heapRelation, AccessShareLock);

	return partitionColumn;
}


/*
 * ColumnNameToVar accepts a relation and column name and returns a Var that
 * represents that column in that relation. If the column doesn't exist or is
 * a system column, an error is thrown.
 */
static Var *
ColumnNameToVar(Relation relation, char *columnName)
{
	Var *partitionColumn = NULL;

	/* Flags for addRangeTableEntryForRelation invocation. */
	const bool useInheritance = false;
	const bool inFromClause = true;

	/*
	 * Flags for addRTEtoQuery invocation. Only need to search column names, so
	 * don't bother adding relation name to parse state.
	 */
	const bool addToJoins = false;
	const bool addToNamespace = false;
	const bool addToVarNamespace = true;

	ParseState *parseState = make_parsestate(NULL);

	RangeTblEntry *rte = addRangeTableEntryForRelation(parseState, relation,
													   NULL, useInheritance,
													   inFromClause);
	addRTEtoQuery(parseState, rte, addToJoins, addToNamespace,
				  addToVarNamespace);

	partitionColumn = (Var *) scanRTEForColumn(parseState, rte, columnName, 0);

	free_parsestate(parseState);

	if (partitionColumn == NULL)
	{
		ereport(ERROR,
				(errcode(ERRCODE_UNDEFINED_COLUMN),
				 errmsg("partition column \"%s\" not found", columnName)));
	}
	else if (!AttrNumberIsForUserDefinedAttr(partitionColumn->varattno))
	{
		ereport(ERROR,
				(errcode(ERRCODE_INVALID_COLUMN_REFERENCE),
				 errmsg("specified partition column \"%s\" is a system column",
						 columnName)));
	}

	return partitionColumn;
}



/*
 * LoadShardRow finds the row for the specified shard identifier in the shard
 * table and copies values from that row into the provided output parameters.
 */
static void
LoadShardRow(int64 shardId, Oid *relationId, char **minValue, char **maxValue)
{
	const int scanKeyCount = 1;

	RangeVar *heapRangeVar = NULL, *indexRangeVar = NULL;
	Relation heapRelation = NULL, indexRelation = NULL;
	IndexScanDesc indexScanDesc = NULL;
	ScanKeyData scanKey[scanKeyCount];
	HeapTuple heapTuple = NULL;

	Datum relationIdDatum = 0;
	Datum minValueDatum = 0;
	Datum maxValueDatum = 0;
	bool isNull = false;

	heapRangeVar = makeRangeVar(METADATA_SCHEMA, SHARD_TABLE_NAME, -1);
	indexRangeVar = makeRangeVar(METADATA_SCHEMA, SHARD_PKEY_IDX, -1);

	heapRelation = relation_openrv(heapRangeVar, AccessShareLock);
	indexRelation = relation_openrv(indexRangeVar, AccessShareLock);

	ScanKeyInit(&scanKey[0], 1, BTEqualStrategyNumber, F_INT8EQ,
				Int64GetDatum(shardId));

	indexScanDesc = index_beginscan(heapRelation, indexRelation, SnapshotNow,
								  scanKeyCount, 0);
	index_rescan(indexScanDesc, scanKey, scanKeyCount, NULL, 0);

	// TODO: Do I need to check scan->xs_recheck and recheck scan key?
	heapTuple = index_getnext(indexScanDesc, ForwardScanDirection);
	if (HeapTupleIsValid(heapTuple))
	{
		TupleDesc tupleDescriptor = RelationGetDescr(heapRelation);

		relationIdDatum = heap_getattr(heapTuple, ATTR_NUM_SHARD_RELATION_ID,
									   tupleDescriptor, &isNull);
		minValueDatum = heap_getattr(heapTuple, ATTR_NUM_SHARD_MIN_VALUE,
									 tupleDescriptor, &isNull);
		maxValueDatum = heap_getattr(heapTuple, ATTR_NUM_SHARD_MAX_VALUE,
									 tupleDescriptor, &isNull);

		/* convert and deep copy row's values */
		(*relationId) = DatumGetObjectId(relationIdDatum);
		(*minValue) = TextDatumGetCString(minValueDatum);
		(*maxValue) = TextDatumGetCString(maxValueDatum);

	}
	else
	{
		ereport(ERROR, (errmsg("could not find entry for shard " INT64_FORMAT,
						shardId)));
	}

	index_endscan(indexScanDesc);
	index_close(indexRelation, AccessShareLock);
	relation_close(heapRelation, AccessShareLock);

	return;
}


/*
 * TupleToPlacement populates a Placement using values from a row of the
 * placements configuration table and returns a pointer to that struct. The
 * input tuple must not contain any NULLs.
 */
static Placement *
TupleToPlacement(HeapTuple heapTuple, TupleDesc tupleDescriptor)
{
	Placement *placement = NULL;
	bool isNull = false;

	Datum idDatum = heap_getattr(heapTuple, ATTR_NUM_PLACEMENT_ID,
								 tupleDescriptor, &isNull);
	Datum shardIdDatum = heap_getattr(heapTuple, ATTR_NUM_PLACEMENT_SHARD_ID,
									  tupleDescriptor, &isNull);
	Datum nodeNameDatum = heap_getattr(heapTuple, ATTR_NUM_PLACEMENT_NODE_NAME,
									   tupleDescriptor, &isNull);
	Datum nodePortDatum = heap_getattr(heapTuple, ATTR_NUM_PLACEMENT_NODE_PORT,
									   tupleDescriptor, &isNull);

	Assert(!HeapTupleHasNulls(heapTuple));

	placement = palloc0(sizeof(Placement));
	placement->id = DatumGetInt64(idDatum);
	placement->shardId = DatumGetInt64(shardIdDatum);
	placement->nodeName = TextDatumGetCString(nodeNameDatum);
	placement->nodePort = DatumGetInt32(nodePortDatum);

	return placement;
}
