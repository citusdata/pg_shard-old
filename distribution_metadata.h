/*-------------------------------------------------------------------------
 *
 * distribution_metadata.h
 *		  Cluster metadata handling for pg_shard
 *
 * Portions Copyright (c) 2014, Citus Data, Inc.
 *
 * IDENTIFICATION
 *		  distribution_metadata.h
 *
 *-------------------------------------------------------------------------
 */

#ifndef PG_SHARD_DISTRIBUTION_METADATA_H
#define PG_SHARD_DISTRIBUTION_METADATA_H

#include "postgres.h"
#include "c.h"
#include "fmgr.h"
#include "postgres_ext.h"

#include "nodes/pg_list.h"
#include "nodes/primnodes.h"


/* Configuration for distributing tables resides in this schema. */
#define METADATA_SCHEMA_NAME "pgs_distribution_metadata"

/* Shard interval information is stored in the shard table. */
#define SHARD_TABLE_NAME "shard"

/* The table has a primary key index for fast lookup. */
#define SHARD_PKEY_INDEX_NAME "shard_pkey"

/* The table has an index to expedite lookup by relation identifier */
#define SHARD_RELATION_INDEX_NAME "shard_relation_index"

/* Semantic names clarify specific tuple attributes from the shard table. */
#define ATTR_NUM_SHARD_ID 1
#define ATTR_NUM_SHARD_RELATION_ID 2
#define ATTR_NUM_SHARD_MIN_VALUE 3
#define ATTR_NUM_SHARD_MAX_VALUE 4

/* Shard placement information is stored in the shard placement table. */
#define SHARD_PLACEMENT_TABLE_NAME "shard_placement"

/* The table has an index to expedite lookup by shard identifier */
#define SHARD_PLACEMENT_SHARD_INDEX_NAME "shard_placement_shard_index"

/* Semantic names explain specific tuple attributes from the shard placement table. */
#define ATTR_NUM_SHARD_PLACEMENT_ID 1
#define ATTR_NUM_SHARD_PLACEMENT_SHARD_ID 2
#define ATTR_NUM_SHARD_PLACEMENT_SHARD_STATE 3
#define ATTR_NUM_SHARD_PLACEMENT_NODE_NAME 4
#define ATTR_NUM_SHARD_PLACEMENT_NODE_PORT 5

/*
 * Partition schemes are stored in the partition table, one for each distributed
 * table tracked by pg_shard.
 */
#define PARTITION_TABLE_NAME "partition"

/* Semantic names clarify specific tuple attributes from the partition table. */
#define ATTR_NUM_PARTITION_RELATION_ID 1
#define ATTR_NUM_PARTITION_KEY 2

/*
 * ShardInterval contains information about a particular shard in a distributed
 * table. ShardIntervals have a unique identifier, a reference back to the table
 * they distribute, and min and max values for the partition column of rows that
 * are contained within the shard (this range is inclusive).
 *
 * All fields are required.
 */
typedef struct ShardInterval
{
	int64 id;			/* unique identifier for the shard */
	Oid relationId;		/* id of the shard's distributed table */
	Datum minValue;		/* a shard's typed min value datum */
	Datum maxValue;		/* a shard's typed max value datum */
	Oid valueTypeId;	/* typeId for minValue and maxValue Datums */
} ShardInterval;


/*
 * ShardPlacement represents information about the placement of a shard in a
 * distributed table. ShardPlacements have a unique identifier, a reference to
 * the shard they place, a textual hostname to identify the host on which the
 * shard resides, and a port number to use when connecting to that host.
 *
 * All fields are required.
 */
typedef struct ShardPlacement
{
	int64 id;		/* unique identifier for the shard placement */
	int64 shardId;	/* identifies shard for this shard placement */
	char *nodeName;	/* hostname of machine hosting this shard */
	int32 nodePort;	/* port number for connecting to host */
} ShardPlacement;


extern List * LoadShardList(Oid distributedTableId);
extern ShardInterval * LoadShardInterval(int64 shardId);
extern List * LoadShardPlacementList(int64 shardId);
extern Var * PartitionColumn(Oid distributedTableId);
extern Datum TestDistributionMetadata(PG_FUNCTION_ARGS);


#endif /* PG_SHARD_DISTRIBUTION_METADATA_H */
