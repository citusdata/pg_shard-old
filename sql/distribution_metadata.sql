-- ===================================================================
-- create test functions
-- ===================================================================

CREATE FUNCTION load_shard_id_array(regclass)
	RETURNS bigint[]
	AS 'pg_shard'
	LANGUAGE C STRICT;

CREATE FUNCTION load_shard_interval_array(bigint)
	RETURNS integer[]
	AS 'pg_shard'
	LANGUAGE C STRICT;

CREATE FUNCTION load_shard_placement_array(bigint)
	RETURNS text[]
	AS 'pg_shard'
	LANGUAGE C STRICT;

CREATE FUNCTION partition_column_id(regclass)
	RETURNS smallint
	AS 'pg_shard'
	LANGUAGE C STRICT;

-- ===================================================================
-- test distribution metadata functionality
-- ===================================================================

-- set up a table and "distribute" it manually
CREATE TABLE events (
	id bigint,
	name text
);

INSERT INTO pgs_distribution_metadata.shard
	(id, relation_id, storage, min_value, max_value)
VALUES
	(1, 'events'::regclass, 't', '0', '10'),
	(2, 'events'::regclass, 't', '10', '20'),
	(3, 'events'::regclass, 't', '20', '30'),
	(4, 'events'::regclass, 't', '30', '40');

INSERT INTO pgs_distribution_metadata.shard_placement
	(id, node_name, node_port, shard_id, shard_state)
VALUES
	(101, 'cluster-worker-01', 5432, 1, 0),
	(102, 'cluster-worker-01', 5432, 2, 0),
	(103, 'cluster-worker-02', 5433, 3, 0),
	(104, 'cluster-worker-02', 5433, 4, 0),
	(105, 'cluster-worker-03', 5434, 1, 0),
	(106, 'cluster-worker-03', 5434, 2, 0),
	(107, 'cluster-worker-04', 5435, 3, 0),
	(108, 'cluster-worker-04', 5435, 4, 0);

INSERT INTO pgs_distribution_metadata.partition (relation_id, partition_method, key)
VALUES
	('events'::regclass, 'h', 'name');

-- should see above shard identifiers
SELECT load_shard_id_array('events');

-- should see empty array (catalog is not distributed)
SELECT load_shard_id_array('pg_type');

-- should see array with first shard range
SELECT load_shard_interval_array(1);

-- should see error for non-existent shard
SELECT load_shard_interval_array(5);

-- should see two placements
SELECT load_shard_placement_array(2);

-- should see error for non-existent shard
SELECT load_shard_placement_array(6);

-- should see column id of 'name'
SELECT partition_column_id('events');

-- should see error (catalog is not distributed)
SELECT partition_column_id('pg_type');

-- clean up after ourselves
DROP TABLE events;
