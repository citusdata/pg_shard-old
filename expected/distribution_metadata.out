-- ===================================================================
-- create test functions
-- ===================================================================
CREATE FUNCTION test_distribution_metadata(oid)
	RETURNS void
	AS 'pg_shard', 'TestDistributionMetadata'
	LANGUAGE C STRICT;
-- ===================================================================
-- test distribution metadata functionality
-- ===================================================================
CREATE TABLE events (
	id bigserial,
	name text
);
INSERT INTO pgs_distribution_metadata.shard
	(relation_id, min_value, max_value)
VALUES
	('events'::regclass, '0', '10'),
	('events'::regclass, '10', '20'),
	('events'::regclass, '20', '30'),
	('events'::regclass, '30', '40');
INSERT INTO pgs_distribution_metadata.placement
	(node_name, node_port, shard_id)
VALUES
	('foo', 123, 1),
	('foo', 123, 2),
	('bar', 456, 3),
	('bar', 456, 4);
INSERT INTO pgs_distribution_metadata.partition_strategy (relation_id, key)
VALUES
	('events'::regclass, 'name');
SELECT test_distribution_metadata('events'::regclass);
INFO:  Table is partitioned using column #2, which is of type "text"
INFO:  Found 4 shards...
INFO:  Shard #4
INFO:  	relation:	events
INFO:  	min value:	30
INFO:  	max value:	40
INFO:  	1 placements:
INFO:  		Placement #4
INFO:  			shard:	4
INFO:  			node name:	bar
INFO:  			node port:	456
INFO:  Shard #3
INFO:  	relation:	events
INFO:  	min value:	20
INFO:  	max value:	30
INFO:  	1 placements:
INFO:  		Placement #3
INFO:  			shard:	3
INFO:  			node name:	bar
INFO:  			node port:	456
INFO:  Shard #2
INFO:  	relation:	events
INFO:  	min value:	10
INFO:  	max value:	20
INFO:  	1 placements:
INFO:  		Placement #2
INFO:  			shard:	2
INFO:  			node name:	foo
INFO:  			node port:	123
INFO:  Shard #1
INFO:  	relation:	events
INFO:  	min value:	0
INFO:  	max value:	10
INFO:  	1 placements:
INFO:  		Placement #1
INFO:  			shard:	1
INFO:  			node name:	foo
INFO:  			node port:	123
 test_distribution_metadata 
----------------------------
 
(1 row)

